File: 2022/14/14.kt
fun main() {
    val lines = generateSequence(::readlnOrNull).toList()
        .map { it.split(" -> ", ",").map { it.toInt() }.chunked(2) }
    val width = lines.flatten().maxOf { it[0] } + 200
    val height = lines.flatten().maxOf { it[1] } + 2
    val field = MutableList(height) { MutableList(width) { '.' } }
    fun range(a: Int, b: Int) = minOf(a, b)..maxOf(a, b)
    fun drawLine(rx: IntRange, ry: IntRange) = rx.map { x -> ry.map { y -> field[y][x] = '#' } }
    lines.map { it.zipWithNext { (x1, y1), (x2, y2) -> drawLine(range(x1, x2), range(y1, y2))} }

    fun fill(): Int {
        var endDfs = false
        fun dfs(x: Int, y: Int): Boolean { 
            endDfs = endDfs || y >= field.size || field[0][500] == 'O'
            if (endDfs || field[y][x] != '.')
                return endDfs
            if (!(dfs(x, y+1) || dfs(x-1, y+1) || dfs(x+1, y+1))) 
                field[y][x] = 'O'
            return true
        }
        while (!endDfs) dfs(500, 0)
        return field.flatten().count { it == 'O' } 
    }
    println(fill())
    field.add(MutableList(width) { '#' })
    println(fill())
}

File: 2022/14/14.py
from sys import stdin
lines = [[list(map(int, c.split(','))) for c in l.strip().split(" -> ")] for l in stdin.readlines()]
width = max(x+200 for x, _ in sum(lines, []))
height = max(y for _, y in sum(lines, []))

def solve(floor='.'):
    field = [['.'] * width for _ in range(height + 2)] + [[floor] * width]
    for walls in lines:
        for (x1, y1), (x2, y2) in zip(walls, walls[1:]):
            for x in range(min(x1, x2), max(x1, x2)+1):
                for y in range(min(y1, y2), max(y1, y2)+1):
                    field[y][x] = '#'

    while True:
        sx, sy = 500, 0
        while sy+1 < len(field):
            for xa in (0, -1, 1):
                if field[sy+1][sx+xa] == '.':
                    sy, sx = sy+1, sx+xa
                    break
            else:
                field[sy][sx] = 'o'
                break
        if sy+1 >= len(field) or field[0][500] == 'o':
            return sum(field, []).count("o")

print(solve(floor='.'))
print(solve(floor='#'))
